tensornetwork/linalg/operations.py

```diff
@@ -0,0 +1,405 @@
+  Returns:
+    The Hermitian conjugated `Tensor`.
+  """
+  return tensor.hconj(perm=perm)
+
+
+def sin(tensor: Tensor) -> Tensor:
+  """
+  Return sin of `Tensor`.
+  Args:
+    Tensor: A Tensor.
+  Returns:
+    Tensor
+  """
+  out_array = tensor.backend.sin(tensor.array)
+  return Tensor(out_array, backend=tensor.backend)
+
+
+def cos(tensor: Tensor) -> Tensor:
+  """
+  Return cos of `Tensor`.
+  Args:
+    Tensor: A Tensor.
+  Returns:
+    Tensor
+  """
+  out_array = tensor.backend.cos(tensor.array)
+  return Tensor(out_array, backend=tensor.backend)
+
+
+def exp(tensor: Tensor) -> Tensor:
+  """
+  Return elementwise exp of `Tensor`.
+  Args:
+    Tensor: A Tensor.
+  Returns:
+    Tensor
+  """
+  out_array = tensor.backend.exp(tensor.array)
+  return Tensor(out_array, backend=tensor.backend)
+
+
+def log(tensor: Tensor) -> Tensor:
+  """
+  Return elementwise natural logarithm of `Tensor`.
+  Args:
+    Tensor: A Tensor.
+  Returns:
+    Tensor
+  """
+  out_array = tensor.backend.log(tensor.array)
+  return Tensor(out_array, backend=tensor.backend)
+
+
+def diagonal(tensor: Tensor, offset: int = 0, axis1: int = -2,
+             axis2: int = -1) -> Tensor:
+  """
+  Extracts the offset'th diagonal from the matrix slice of tensor indexed
+  by (axis1, axis2).
+
+  Args:
+    tensor: A Tensor.
+    offset: Offset of the diagonal from the main diagonal.
+    axis1, axis2: Indices of the matrix slice to extract from.
+
+  Returns:
+    out  : A 1D Tensor storing the elements of the selected diagonal.
+  """
+  backend = tensor.backend
+  result = backend.diagonal(tensor.array, offset=offset, axis1=axis1,
+                            axis2=axis2)
+  return Tensor(result, backend=backend)
+
+
+def diagflat(tensor: Tensor, k: int = 0) -> Tensor:
+  """
+  Flattens tensor and places its elements at the k'th diagonal of a new
+  (tensor.size + k, tensor.size + k) `Tensor` of zeros.
+
+  Args:
+    tensor: A Tensor.
+    k     : The elements of tensor will be stored at this diagonal.
+  Returns:
+    out   : A (tensor.size + k, tensor.size + k) `Tensor` with the elements
+            of tensor on its kth diagonal.
+  """
+  backend = tensor.backend
+  result = backend.diagflat(tensor.array, k=k)
+  return Tensor(result, backend=backend)
+
+
+def trace(tensor: Tensor, offset: int = 0, axis1: int = -2,
+          axis2: int = -1) -> Tensor:
+  """Calculate the sum along diagonal entries of the given Tensor. The
+     entries of the offset`th diagonal of the matrix slice of tensor indexed by
+     (axis1, axis2) are summed.
+
+  Args:
+    tensor: A Tensor.
+    offset: Offset of the diagonal from the main diagonal.
+    axis1, axis2: Indices of the matrix slice to extract from.
+
+  Returns:
+    out: The trace.
+  """
+  backend = tensor.backend
+  result = backend.trace(tensor.array, offset=offset, axis1=axis1,
+                         axis2=axis2)
+  return Tensor(result, backend=backend)
+
+
+def ncon(
```

Comment:
TBH, I'd like to veto that `ncon` only accepts `Tensor`. I use `ncon` so much for all kinds of things that it would be annoying to be forced to use it on `Tensor` objects only. I would prefer to modify `ncon` in `ncon_interface.py` instead

Reply:
I'd suggest that tn.ncon accept Tensor and backend.ncon accept the appropriate backend object

Reply:
backends don't have `ncon` implemented. Rather, `ncon` uses backend functions in its implementation

Reply:
It seems a bit odd to me that this is the only function which behaves this way. Is there a principled reason for it to?

Reply:
I think I'd be OK with this, but then I'd like to get rid of one of the two `ncon` functions. Otherwise this will be pretty confusing

Reply:
Will this work with `BlockSparseTensor`s (i.e. `SymmetricBackend`)?

Reply:
It basically just calls ncon on the backend object, so I think it should be fine